src/Features/Core/Portable/ForeachToFor/AbstractForeachToForCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,346 @@
+                if (indexer != null && semanticFact.IsAssignableTo(indexer.ReturnType, foreachType, model.Compilation))
+                {
+                    explicitCast = null;
+                    countName = Count;
+                    return;
+                }
+            }
+
+            // check regular cases (implicitly implemented)
+            string explicitInterface = null;
+            foreach (var current in collectionType.AllInterfaces)
+            {
+                if (!knownCollectionInterfaces.Contains(current.OriginalDefinition))
+                {
+                    continue;
+                }
+
+                // see how the type implements the interface
+                var countSymbol = GetInterfaceMembers(current, get_Count);
+                var indexerSymbol = GetInterfaceMembers(current, get_Item);
+                if (countSymbol == null || indexerSymbol == null)
+                {
+                    continue;
+                }
+
+                var countImpl = collectionType.FindImplementationForInterfaceMember(countSymbol) as IMethodSymbol;
+                var indexerImpl = collectionType.FindImplementationForInterfaceMember(indexerSymbol) as IMethodSymbol;
+                if (countImpl == null || indexerImpl == null)
+                {
+                    continue;
+                }
+
+                if (!semanticFact.IsAssignableTo(indexerImpl.ReturnType, foreachType, model.Compilation))
+                {
+                    continue;
+                }
+
+                // implicitly implemented!
+                if (countImpl.ExplicitInterfaceImplementations.IsEmpty &&
+                    indexerImpl.ExplicitInterfaceImplementations.IsEmpty)
+                {
+                    explicitCast = null;
+                    countName = Count;
+                    return;
+                }
+
+                if (explicitInterface == null)
+                {
+                    explicitInterface = current.ToDisplayString();
+                }
+            }
+
+            // okay, we don't have implicitly implemented one, but we do have explicitly implemented one
+            if (explicitInterface != null)
+            {
+                explicitCast = explicitInterface;
+                countName = Count;
+            }
+        }
+
+        private static bool IsNullOrErrorType(ITypeSymbol type)
+        {
+            return type == null || type is IErrorTypeSymbol;
+        }
+
+        private static IMethodSymbol GetInterfaceMembers(ITypeSymbol interfaceType, string memberName)
+        {
+            var members = interfaceType.GetMembers(memberName);
+            if (!members.IsEmpty)
+            {
+                return (IMethodSymbol)members[0];
+            }
+
+            foreach (var current in interfaceType.AllInterfaces)
+            {
+                var member = GetInterfaceMembers(current, memberName);
```

Comment:
this is an exponential algorithm.  interfaceType.AllInterfaces already walks the entire hierarchy giving you everything.  By walking the hierarchy and then calling GetInterfaceMembers on each interface, you're producing those whole hierarchies over and over again.

Reply:
ah. yes. should be just Interfaces. not allInterfaces.